File: src/main/java/com/cuupa/mailprocessor/delegates/SemanticDelegate.java
package com.cuupa.mailprocessor.delegates;

import com.cuupa.mailprocessor.process.ProcessInstanceHandler;
import com.cuupa.mailprocessor.services.ExternSemanticService;
import com.cuupa.mailprocessor.services.Metadata;
import com.cuupa.mailprocessor.services.SemanticResult;
import org.apache.commons.collections4.CollectionUtils;
import org.camunda.bpm.engine.delegate.DelegateExecution;
import org.camunda.bpm.engine.delegate.JavaDelegate;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public class SemanticDelegate implements JavaDelegate {

    private final ExternSemanticService externSemanticService;

    public SemanticDelegate(final ExternSemanticService externSemanticService) {
        this.externSemanticService = externSemanticService;
    }

    @Override
    public void execute(DelegateExecution delegateExecution) throws Exception {
        ProcessInstanceHandler handler = new ProcessInstanceHandler(delegateExecution);
        final List<SemanticResult>
                semanticResult =
                externSemanticService.getSemanticResult(String.join("", handler.getPlainText()));

        semanticResult.forEach(result -> {
            final Collection<Metadata> metadata = CollectionUtils.emptyIfNull(result.getMetaData());
            handler.addTopic(result.getTopicName())
                   .setSender(result.getSender())
                   .addMetaData(metadata.stream()
                                        .filter(e -> !"sender".equals(e.getName()))
                                        .collect(Collectors.toList()));
        });
    }
}

File: src/main/java/com/cuupa/mailprocessor/delegates/SemanticDelegate.kt
package com.cuupa.mailprocessor.delegates

import com.cuupa.mailprocessor.process.ProcessInstanceHandler
import com.cuupa.mailprocessor.services.semantic.ExternSemanticService
import com.cuupa.mailprocessor.services.semantic.Metadata
import com.cuupa.mailprocessor.services.semantic.SemanticResult
import org.camunda.bpm.engine.delegate.DelegateExecution
import org.camunda.bpm.engine.delegate.JavaDelegate

class SemanticDelegate(private val externSemanticService: ExternSemanticService) : JavaDelegate {

    override fun execute(delegateExecution: DelegateExecution) {
        val handler = ProcessInstanceHandler(delegateExecution)
        externSemanticService.getSemanticResult(getText(handler)).forEach {
            handler.addTopic(it.topicName)
            handler.sender = getSender(it)
            handler.addMetaData(it.metaData.filter(isNotSenderMetadata()))
        }
    }

    private fun getSender(semanticResult: SemanticResult): String? {
        return when (semanticResult.sender) {
            unknown -> semanticResult.metaData.firstOrNull(isSenderMetadata())?.value ?: unknown
            else -> semanticResult.sender
        }
    }

    private fun getText(handler: ProcessInstanceHandler) = handler.plainText.joinToString(" ", "", "")
    private fun isSenderMetadata(): (Metadata) -> Boolean = { "sender" == it.name }
    private fun isNotSenderMetadata(): (Metadata) -> Boolean = { "sender" != it.name }

    companion object {
        private const val unknown = "UNKNOWN"
    }

}